### PR TITLE
Add variables to the grammar

### PR DESCRIPTION
**Pull request by tsandall**
The substantive changes are to the grammar file and the new tests in test_parser.go.

For review purposes, the third commit is the most relevant. The first two are quite self explanatory. The third commit contains a bit of refactoring which I did along the way.

I am going to try refactoring the AST to see if using separate structs for the different kinds of Term is worthwhile. I'll submit another PR for that. In the mean time, it would not hurt to review this.

### CODE REVIEW

File: opalog/opalog.peg [hunk 1]
```diff
@@ -69,10 +82,6 @@ Number <- '-'? Integer ( '.' DecimalDigit+ )? Exponent? {
     return t, err
```
Comment (tim-styra) [inline opalog/opalog.peg]: If JSON doesn't define a limit on the size of a number, then to be truly JSON compliant, we need to handle BigNums.  Does Go's ParseFloat do that?  If not, I'd imagine we should leave a TODO.

File: opalog/opalog.peg [hunk 2]
```diff
@@ -1,64 +1,77 @@
...
-    res := NewKeyValueSet()
-    res.Add(NewKeyValue(valsSl[0].(*Term), valsSl[4].(*Term)))
-    for _, v := range restSl {
-        vSl := toIfaceSlice(v)
-        res.Add(NewKeyValue(vSl[2].(*Term), vSl[6].(*Term)))
+
+    // Non-empty object, first key/value pair.
+    headSlice := head.([]interface{})
+    set.Add(NewKeyValue(headSlice[0].(*Term), headSlice[len(headSlice) - 1].(*Term)))
+
```
Comment (tim-styra) [inline opalog/opalog.peg]: Would be helpful to have a comment that shows an example of what 'head' looks like.  I was confused as to why it wasn't headSlice[0] and headSlice[1].

File: opalog/parser_test.go [hunk 3]
```diff
@@ -6,46 +6,99 @@ package opalog
...
+	assertParseOneTermFail(t, "non-var", "foo-bar")
+	assertParseOneTermFail(t, "non-var2", "foo-7")
+}
+
+func TestObjectWithScalars(t *testing.T) {
+	assertParseOneTerm(t, "number", "{\"abc\": 7, \"def\": 8}", reflectTerm(map[string]int{"abc": 7, "def": 8}))
+	assertParseOneTerm(t, "bool", "{\"abc\": false, \"def\": true}", reflectTerm(map[string]bool{"abc": false, "def": true}))
+	assertParseOneTerm(t, "string", "{\"abc\": \"foo\", \"def\": \"bar\"}", reflectTerm(map[string]string{"abc": "foo", "def": "bar"}))
+	assertParseOneTerm(t, "mixed", "{\"abc\": 7, \"def\": null}", reflectTerm(map[string]interface{}{"abc": 7, "def": nil}))
+}
```
Comment (tim-styra) [inline opalog/parser_test.go]: What do you think about supporting objects where the key is a number or a boolean?  I know the spec doesn't allow it, but if other json parsers handle it, we probably should too.

File: opalog/parser_test.go [hunk 4]
```diff
@@ -79,39 +132,64 @@ func testParse1TermFail(t *testing.T, msg string, expr string) {
...
+		length := xval.Len()
+		arr := make([]*Term, length)
+		for i := 0; i < length; i++ {
+			arr[i] = reflectTerm(xval.Index(i).Interface())
+		}
+		val = arr
+		typ = ARRAY
+	default:
+		panic(fmt.Sprintf("Unexpected type of term: %v", x))
+	}
```
Comment (tim-styra) [inline opalog/parser_test.go]: Feel free to delay this to another commit, but we're not testing combinations of embedded composites, such as [[1,foo,3], {"abc": bar}].

**Comment by tim-styra** [comment]
Overall looks good.  Just a couple thoughts.

I like the code for the parser--very clean and consistent.  I like the refactoring of the test code.  Sorry it was such a mess before.

File: opalog/opalog.peg (hunk 1 above)
Comment (tsandall) [inline opalog/opalog.peg]: Sorry, the commit message was a bit confusing. The JSON spec says the limit is *defined by the implementation*. The spec recommends IEEE-754 double precision numbers (which is what float64 is in Go) for good interoperability. I agree that adding support for arbitrary precision is not needed right now.

File: opalog/opalog.peg (hunk 2 above)
Comment (tsandall) [inline opalog/opalog.peg]: Agreed, will add a comment.

File: opalog/parser_test.go (hunk 3 above)
Comment (tsandall) [inline opalog/parser_test.go]: From the OPA/Opalog perspective, I agree. The Go JSON package will not support this well though, so I think we'll end up discussing this further. For now I will update the tests as you described.

File: opalog/parser_test.go (hunk 4 above)
Comment (tsandall) [inline opalog/parser_test.go]: This should be handled already so I will update the tests.

**Comment by tsandall** [comment]
Thanks for the reviewing the test coverage. I will update the coverage and then merge this in.

**Pull request merged**